libtrellis/tools/ecpmulti.cpp

```diff
@@ -0,0 +1,171 @@
+
+    po::options_description options("Allowed options");
+    options.add_options()("help,h", "show help");
+    options.add_options()("verbose,v", "verbose output");
+    options.add_options()("db", po::value<std::string>(), "Trellis database folder location");
+    options.add_options()("flashsize", po::value<std::uint32_t>()->required(), "Flash size in Mbits, e.g. 2, 4, 8, ..., 128");
+    options.add_options()("input", po::value<std::vector<std::string>>()->required(), "input bitstream file 0..N");
+    options.add_options()("address", po::value<std::vector<std::string>>()->required(), "address to place next bitstream at");
+    options.add_options()("golden", po::value<std::string>(), "input bitstream of a golden bitstream");
+
+    po::positional_options_description pos;
+    options.add_options()("output", po::value<std::string>()->required(), "output bitstream file");
+    pos.add("output", 1);
+
+
+    po::variables_map vm;
+    try {
+        po::parsed_options parsed = po::command_line_parser(argc, argv).options(options).positional(pos).run();
+        po::store(parsed, vm);
+        po::notify(vm);
+    }
+    catch (std::exception &e) {
+        cerr << e.what() << endl << endl;
+        goto help;
+    }
+
+    if (vm.count("help")) {
+        help:
+        cerr << "Project Trellis - Open Source Tools for ECP5 FPGAs" << endl;
+        cerr << "ecpmulti: ECP5 multiboot bitstream assembler" << endl;
+        cerr << endl;
+        cerr << "Copyright (C) 2018 Jens Andersen <jens.andersen@gmail.com>" << endl;
+        cerr << endl;
+        cerr << options << endl;
+        return vm.count("help") ? 0 : 1;
+    }
+
+    if (vm.count("flashsize")) {
+        uint32_t flash_size_mbit = vm["flashsize"].as<uint32_t>();
+        flash_size_bytes = flash_size_mbit * 1024* 1024 / 8;
+        cout << "Using flashsize " << flash_size_mbit << "(" << flash_size_bytes << " bytes)" << endl;
+    }
+
+    auto inputs = vm.at("input").as<std::vector<std::string>>();
+    auto addresses = vm.at("address").as<std::vector<std::string>>();
+
+    if (inputs.size() != (addresses.size() + 1)) {
+        cerr << "Inputs " << inputs.size() << " offsets " << addresses.size() << endl;
+        cerr << "There must be one address specified for each input file" << endl;
+        return 1;
+    }
+
+    if (vm.count("db")) {
+        database_folder = vm["db"].as<string>();
+    }
+
+    try {
+        load_database(database_folder);
+    } catch (runtime_error &e) {
+        cerr << "Failed to load Trellis database: " << e.what() << endl;
+        return 1;
+    }
+
+    ofstream out_file(vm["output"].as<string>(), ofstream::trunc);
+    if (!out_file) {
+        cerr << "Failed to open output file" << endl;
+        return 1;
+    }
+    uint32_t idcode = 0;
+    uint32_t nextaddr = 0;
+    const map<string, string> bs_options = {{"multiboot", "yes"} };
+    for(uint32_t i = 0; i < inputs.size(); i++) {
+        string filename = inputs[i];
```

Comment:
I tend to prefer `.at(i)` rather than `[i]` for all but the most performance-critical of paths.